rust/src/quic/frames.rs

```diff
@@ -115,13 +122,124 @@ impl fmt::Display for StreamTag {
     }
 }
 
+#[derive(Debug, PartialEq)]
+pub(crate) struct Ack {
+    pub largest_acknowledged: u64,
+    pub ack_delay: u64,
+    pub ack_range_count: u64,
+    pub first_ack_range: u64,
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) struct Crypto {
+    pub ciphers: Vec<TlsCipherSuiteID>,
+    pub extv: Vec<QuicTlsExtension>,
+    //does not work, because of lifetime due to references to the slice used for parsing
+    //and cannot manage to create a stucture with the Vector the slices refer to...
+    //pub exts: Vec<TlsExtension>,
+}
+
 #[derive(Debug, PartialEq)]
 pub(crate) enum Frame {
     Padding,
+    Ack(Ack),
+    Crypto(Crypto),
     Stream(Stream),
     Unknown(Vec<u8>),
 }
 
+fn parse_ack_frame(input: &[u8]) -> IResult<&[u8], Frame, QuicError> {
+    let rest = input;
+    let (rest, largest_acknowledged) = quic_var_uint(rest)?;
+    let (rest, ack_delay) = quic_var_uint(rest)?;
+    let (rest, ack_range_count) = quic_var_uint(rest)?;
+    let (rest, first_ack_range) = quic_var_uint(rest)?;
+
+    if ack_range_count != 0 {
+        //TODO RFC9000 section 19.3.1.  ACK Ranges
+        return Err(nom::Err::Error(QuicError::UnhandledYet));
+    }
+
+    Ok((
+        rest,
+        Frame::Ack(Ack {
+            largest_acknowledged,
+            ack_delay,
+            ack_range_count,
+            first_ack_range,
+        }),
+    ))
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct QuicTlsExtension {
+    pub etype: TlsExtensionType,
+    pub values: Vec<u8>,
+    pub indices: Vec<usize>,
```

Comment:
This felt better than having `Vec<Vec<u8>>`
Not sure it is better...

Reply:
I think I prefer Vec<Vec<u8>> over having 2 separate Vecs that needs to be kept in sync

Reply:
Will do unless Pierre finds a better way to store the tls extensions...

Reply:
> This felt better than having `Vec<Vec<u8>>` Not sure it is better...

I'm not sure if it helps here, but something like:
```
pub values: Vec<(usize, u8>),
```
should be possible as well.

Reply:
I do not want a vector of tuples, both vectors do not have the same length ;-)

Reply:
Note: this is also related to your other comment on how to store extensions.

What are you trying to store here? I do not get why, you seem to create a slice of binary data with indices, but that look like a complex (and not rust-idiomatic at all) way to store things.
Do this need to be passed to C? Are there any constraints I missed here?

Otherwise, the rust way would be to define an enum (you don't need to store the type, since it is implicit in the enum variant).
for ex:
```rust
pub enum QuicTlsExtension {
   // example using a struct-like variant
   SNI{ sni_list: Vec<String> },
   // or, using anonymous variants
   SLPN(Vec<String>)
}
```
Then, from the parsed TLS extension, you'd just have to `map` values to own them, and add the variant constructor.

Reply:
> Then, from the parsed TLS extension, you'd just have to map values to own them, and add the variant constructor.

My question is : There is no other way than map the 28 variants of the tls_parser enum and add the variant constructor ?
I want to own them directly if possible...
I want to store a `Vec<tls_parser::TlsExtension>` into my transaction, without needing to recreate a `QuicTlsExtension`

Reply:
Not that I know of :/
One way to do that would be to use `Cow<'a, &[u8]>` everywhere in `TlsExtension`, but that would require a breaking change in that create. Can you submit a feature request issue to [`tls-parser`](https://github.com/rusticata/tls-parser)?

Reply:
> I think I prefer Vec<Vec> over having 2 separate Vecs that needs to be kept in sync

Done this for the moment

Maybe `Cow` is good...